package/kernelctl/kernelctl

```diff
@@ -0,0 +1,191 @@
+		${bf}    help                    ${sf}Show this help message.
+		${bf}    list                    ${sf}List available kernels.
+		${bf}    show                    ${sf}Show the current configured kernel.
+		${bf}    delete <kernel>         ${sf}Delete kernel from the staging dir. WARNING this is irreversible.
+		${bf}    set <kernel>            ${sf}Change booting kernel.
+
+		${bf}        <kernel>            ${sf}Kernel name or number (from 'list' command) or "default" to revert to the upstram kernel.
+	EOM
+    echo -e "$msg"
+}
+
+# backup current kernel
+backup() {
+    if [[ "$1" == "vendor" ]]; then
+        archive_name="vendor-$(< /etc/version)"
+    else
+        archive_name=$1
+    fi
+
+    archive_file="$kernelctl_dir"/"$archive_name".tar.bz2
+
+    if [[ -e "$archive_file" ]]; then
+        read -r -d "" msg <<- EOM
+			It looks like we already have a backup copy of $archive_name.
+			If you really want to back it up again please run
+
+			$(basename "$0") delete $archive_name
+			$(basename "$0") backup $archive_name
+		EOM
+        echo -e "$msg"
+    else
+        tar cpjf "$archive_file" lib/modules/* boot/zImage* boot/*.dtb
+    fi
+    echo "$archive_name" > "$kernelctl_status"
+}
+
+# get available kernels
+get_kernels() {
+    mapfile -t kernels < <(find "$kernelctl_dir" -path "*.tar.bz2" -print0 | xargs -0 -I"{}" basename {} .tar.bz2)
+}
+
+# translate input into a kernel name
+to_kernel_name() {
+    local ker
+    get_kernels
+    if ((0 < $1 && $1 <= ${#kernels[@]})); then
+        echo "${kernels[$(($1 - 1))]}"
+        return
+    elif [[ "$1" = "default" ]]; then
+        ker="vendor-$(< /etc/version)"
+    else
+        ker="$1"
+    fi
+    if [[ $(echo "${kernels[@]}" | grep -ow "$ker" | wc -w) -gt 0 ]]; then
+        echo "$ker"
+    else
+        echo "Can't find $ker in the staging area"
+        exit 1
+    fi
+}
+
+# list available kernels
+list() {
+    local current_kernel
+    echo -e "${c1}Available kernels:${sf}"
+    get_kernels
+    current_kernel="$(< $kernelctl_status)"
+    for i in "${!kernels[@]}"; do
+        if [[ "$current_kernel" = "${kernels[$i]}" ]]; then cur=" ${c2}*${sf}"; else cur=""; fi
+        echo -e "  ${bf}[$((i + 1))]${sf}\t${kernels[$i]}$cur"
+    done
+}
+
+# show the current configured kernel
+show() {
+    echo -e "${c1}Current kernel:${sf}"
+    echo -e "  ${bf}$(< $kernelctl_status)${sf}"
+}
+
+# change the kernel that will boot next time
+set() {
+    new_kernel_name=$(to_kernel_name "$1")
+    new_kernel="$kernelctl_dir/${new_kernel_name}.tar.bz2"
+    current_kernel="$kernelctl_dir/$(< $kernelctl_status).tar.bz2"
+    if [[ ! -e "$current_kernel" ]]; then
+        rear -r -d '' msg <<- EOM
+			${bf}There is no copy of the current running kernel in the staging
+			area, you might have accidentally deleted it!${sf}
+			Try making a backup first
+		EOM
+        echo -e "$msg"
+        exit 1
+    fi
+    tar tjf "$current_kernel" | sort -r | xargs rm -d 2> /dev/null
```

Comment:
Do we care if `tar tjf "$current_kernel"` doesn't return any values?
```suggestion
    tar tjf "$current_kernel" | sort -r | xargs -r rm -d 2> /dev/null
```

Reply:
Also, what does `rm -d` do? I'm not seeing it in any help manual/output.

Reply:
> Do we care if `tar tjf "$current_kernel"` doesn't return any values?

No but I guess an extra safety will not hurt. A am also considering removing `2> /dev/null`, this will help to identify badly formatted tarballs when reviewing packages.

Reply:
> Also, what does `rm -d` do? I'm not seeing it in any help [manual](https://www.busybox.net/downloads/BusyBox.html)/output.

It removes empty directories but not full one.